### DISCUSSION TITLE
Keyboard and Mouse Only Mode

### DISCUSSION THREAD

**Discussion by kj4ezj** [rustdesk/rustdesk#7926] [Ideas]
Hey RustDesk community! Thanks for an amazing, open system that enabled me to ditch TeamViewer a number of years ago.

## Background
I have a media center "PC" (running Linux) connected to my TV which I normally control with a little handheld bluetooth keyboard and mouse combo, but I lost the controller while traveling. While I wait on another, I naturally turned to RustDesk to control my computer across the room using my Android. RustDesk is great for remote desktop, but it is clunky for controlling YouTube or other video playback, to say the least.

> [!NOTE]
> > I can think of a number of other use-cases for this mode, such as controlling a presentation with one's phone, sharing a computer for multiplayer "flash" games, or controlling a server which has a display but not a whole crash cart.

## Feature Request
**I would like to collaborate to add a connection mode to RustDesk that only sends keyboard and mouse input.**

I imagine a connection option right below `Connect ${RUSTDESK_ID}` like `Keyboard + Mouse only` or `KM only`.

![RustDesk-annotated](https://github.com/rustdesk/rustdesk/assets/34947245/154faf6f-5639-4530-b664-a1a6f8f7ad45)

When you click it, you authenticate like normal and then the RustDesk client drops you on a screen with your Android keyboard, arrow keys, a toolbar of ancillary keys (`Ctrl`, `Alt`, etc.), maybe copy and paste buttons, and the rest of the screen real estate behaves just like a laptop touchpad. No video or display is streamed to the client.

> [!TIP]
> > Behaving like a laptop touchpad means dragging one finger around on the client moves the cursor around on the target computer's screen, one finger tap performs a left-click, double tap performs a double-click, a two-finger tap performs a right-click, dragging two fingers horizontally or vertically on the client scrolls in the respective direction on the target computer, perhaps a three-finger tap performs a middle-click, etc. We will also need to provide a mechanism to click and drag, for example to select text, such as one-finger double-tap with a hold and drag on the second tap.

On mobile clients, there must be some mechanism to pull the on-screen (Android) keyboard back up if it is dismissed.

Physical keybords, mice, or touchpads attached to the client must also be able to be captured on behalf of the target computer, with some escape mechanism to return control to the client computer.

Usage instructions can be printed in the screen real estate used for touch input.

## Next Steps
More ideas.
1. Bonus points if the arrow keys are a bit larger, positioned on the lower right-hand side of the screen just above the Android keyboard and in a common orientation such that they are easy to press with the thumb to use for games or to seek in video playback.
1. A joystick in the lower left-hand side of the screen for WASD would be cool, also with games in mind.
1. I can also imagine a special presentation mode for PowerPoint or whatever. Perhaps arrow keys, F-keys to enter and exit presentation mode in PowerPoint on the target computer, a play/pause button, and maybe that is it.
1. It would be cool to be able to lock your screen and use the volume rocker to go forwards and backwards in a presentation.
1. A media mode might be cool, with a touchpad and common media controls.
1. A macro mode would be cool where you can send predetermined text or input similar to DuckyScript.
1. Maybe you can swipe left or right outside of the touchpad area to cycle through all of these various views I have described.
1. Being able to switch back-and-forth between this KM-input mode, full remote desktop, or a file transfer without having to re-authenticate could be nice.

## Notes
I understand the existing remote desktop connection has a mechanism for touch or mouse input, but it does not behave as I described above and there is no mechanism to stop streaming the screen as far as I can tell. It is just clunky for this use-case.

I welcome suggestions of other FOSS apps compatible with Linux and Android that may satisfy my use-case. However, if these exist, my feature request still stands. We all have RustDesk installed, we have accepted the security implications of using it, and RustDesk is remote control software that I believe should service these use-cases for computer remote control.

> [!NOTE]
> > I am a software engineer with a strong background in programming. I think it would be fun to help implement this, but I am new to Rust and completely unfamiliar with the RustDesk codebase on top of my own time constraints I am sure all other open source contributors understand, so it would be great to pair up with someone more experienced in this project who could help point me in the right direction to get me started if this is a feature the community is interested in adding.

Thanks for taking the time to read my suggestions, and I look forward to any discussion.

**Comment by seahindeniz** [comment]
I would love to see this controller feature for the mobile client as well. Sometimes, it is good to control a PC/device from a certain distance only for media needs, such as play, pause, stop, next, previous, volume up/down, and a few other extra keys from the keyboard (possibly customizable).

**Reply by RoninTech** [reply]
Or travel without a keyboard/mouse and just use RustDesk on your phone.  Like the old [Valence](https://cafbit.com/valence/) android app.  Not for writing a novel with but fine for many things.

**Comment by xlionjuan** [comment]
If you're using Android and your computer has Bluetooth, you could try

https://play.google.com/store/apps/details?id=io.appground.blek

**Reply by seahindeniz** [reply]
Thanks for the suggestion, I have tried using this but after connecting 1 time, the Bluetooth doesn't seem to work anymore. I believe my only option to have this

**Reply by xlionjuan** [reply]
This kind of tool confuses the system sometimes, just retry or quit app and retry may fix the problem

**Reply by seahindeniz** [reply]
Already tried, even cleared cache for Bluetooth's system app. Didn't work, unfortunately. 
Though I found this app which works over wi-fi connection but requires a server https://play.google.com/store/apps/details?id=com.Relmtech.Remote

**Comment by cczw123** [comment]
I am happy to take a look and work on it.

**Comment by trymeouteh** [comment]
I would like to see an option for the host to disable screen sharing and sharing its audio to the client device. This will be a useful security settings on the host.

And I would like to see an option on the client to disable/hide the shared screen and any audio from the host device. This will be a useful setting on the client to reduce bandwidth usage by not requesting the video or audio feed from the device,

I would also like to see a setting for the host device to not popup a "controller" window that appears when you connect a client to the host which always unfocuses the current window. This is not ideal for when using Rustdesk as a remote.
